presto-system-access-control/src/main/java/com/facebook/presto/security/KerberosExactMatchAccessControl.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.security;
+
+import com.facebook.presto.spi.CatalogSchemaName;
+import com.facebook.presto.spi.CatalogSchemaTableName;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.security.Identity;
+import com.facebook.presto.spi.security.Privilege;
+import com.facebook.presto.spi.security.SystemAccessControl;
+
+import javax.security.auth.kerberos.KerberosPrincipal;
+
+import java.security.Principal;
+
+import static com.facebook.presto.spi.StandardErrorCode.PERMISSION_DENIED;
+
+public class KerberosExactMatchAccessControl
+        implements SystemAccessControl
+{
+    @Override
+    public void checkCanSetUser(Principal principal, String userName)
+    {
+        // Allow unauthenticated access.
+        if (principal == null) {
+            return;
+        }
+
+        /*
+         * This is almost certainly a configuration issue. If people are authenticating with e.g. LDAP,
+         * this is not the SystemAccessControl plugin you should be using.
+         */
+        if (!(principal instanceof KerberosPrincipal)) {
+            throw new PrestoException(
+                    PERMISSION_DENIED,
+                    "The 'kerberos-exact-match' SystemAccessControl plugin only works with " +
+                    "Kerberos authentication. Contact you Presto administrator.");
+        }
+
+        KerberosPrincipal kerberosPrincipal = (KerberosPrincipal) principal;
+
+        String realmName = kerberosPrincipal.getRealm();
+        String kerberosUserName = kerberosPrincipal.getName();
+
+        if (realmName != null && realmName.length() != 0) {
+            kerberosUserName = kerberosUserName.substring(0, kerberosUserName.length() - realmName.length() - 1);
+        }
+
+        if (!userName.equals(kerberosUserName)) {
+            throw new PrestoException(PERMISSION_DENIED, String.format("Principal %s may not set user %s", principal.getName(), userName));
+        }
+    }
+
+    @Override
+    public void checkCanSetSystemSessionProperty(Identity identity, String propertyName)
```

Comment:
Counter-proposal: Allow multiple SystemAccessControl plugins, each of which only implements the methods they care about and doesn't throw for the others. That way you can mix e.g. this with a (making stuff up right now) OnlyAdminCanDropTableSystemAccessControl.

Checking if something is allowed goes as follows:

```
public class MasterSystemAccessControl
List<SystemAccessControl> modules;
public void checkCanDropTable(...)
{
    for (SAC module : modules) {
        canCheckDropTable(...)
    }
}
```

I can write this up in an issue if this sounds reasonable to you both (or merits extensive further discussion outside of this PR)

@dain as well.